File: testing_ex/filas_java/FilaSeq.java
package filas;

public class FilaSeq {
	private int dados[];
	private int inicio;
	private int fim;
	private int nElementos;
	//private int tamMax;
	
	public FilaSeq() {
		inicio = 0;
		fim = -1;
		nElementos = 0;
		
		dados =  new int[100];
	}
	
	public FilaSeq(int n) {
		inicio = 0;
		fim = -1;
		nElementos = 0;
		
		dados =  new int[n];
	}

	/** Verifica se a Fila está vazia */
	public boolean vazia () {
		if (nElementos == 0)
			return true;
		else
			return false;
	}

	/**Verifica se a Fila está cheia */
	public boolean cheia () {
		if (nElementos == dados.length)
			return true;
		else
			return false;
	}

	/** Obtém o tamanho da Fila */
	public int tamanho() {
		return nElementos;
	}

	/** Consulta o elemento do início da fila.
	    Retorna -1 se a fila estiver vazia. */
	public int primeiro() {
		if (vazia())
			return -1; // Erro: Fila vazia 
		
		return dados[inicio];
	}

	/**Insere um elemento no fim de uma fila
    Retorna false se a fila estiver cheia, true caso contrário. */
	public boolean insere(int valor) {
		if (cheia()){
			return false;
		}
	
		fim = (fim + 1) % dados.length; // Circularidade 
	    dados[fim] = valor;
		nElementos++;
		return true;
	}

	/**Remove o elemento do início da fila e retorna o valor removido.
	    Retorna -1 se a fila estiver vazia.*/
	public int remove() {
		if (vazia())
			return -1;
	
		// Guarda o valor a ser removido
		//int valor = primeiro();
		int valor = dados[inicio];
		
		// Atualiza o atributo inicio;
		inicio = (inicio + 1) % dados.length; //Circularidade 
		nElementos--;
		return valor;
	}

}

File: FS/FilaSeq.py
class FilaSeq:
    def __init__(self, TAM_MAX):
        # Inicialização da fila com tamanho máximo
        self.TAM_MAX = TAM_MAX
        self.dados = [0] * TAM_MAX
        self.inicio = 0
        self.fim = -1
        self.nElementos = 0

    def vazia(self):
        # Verifica se a fila está vazia
        return self.nElementos == 0

    def cheia(self):
        # Verifica se a fila está cheia
        return self.nElementos == self.TAM_MAX

    def tamanho(self):
        # Retorna o número de elementos na fila
        return self.nElementos

    def primeiro(self):
        # Retorna o primeiro elemento da fila, se não estiver vazia
        if self.vazia():
            return -1
        return self.dados[self.inicio]

    def insere(self, valor):
        # Insere um valor na fila
        if self.cheia():
            return False  # Retorna False se a fila estiver cheia
        self.fim = (self.fim + 1) % self.TAM_MAX  # Atualiza o índice de fim (com tratamento de circularidade)
        self.dados[self.fim] = valor
        self.nElementos += 1  # Incrementa o número de elementos na fila
        return True

    def remove(self):
        # Remove o primeiro elemento da fila
        if self.vazia():
            return -1  # Retorna -1 se a fila estiver vazia
        valor = self.dados[self.inicio]  # Valor a ser removido
        self.inicio = (self.inicio + 1) % self.TAM_MAX  # Atualiza o índice de início (com tratamento de circularidade)
        self.nElementos -= 1  # Decrementa o número de elementos na fila
        return valor  # Retorna o valor removido
